osu.Game/Online/Chat/ChannelManager.cs

```diff
@@ -408,13 +419,72 @@ public void LeaveChannel(Channel channel)
 
             joinedChannels.Remove(channel);
 
+            // Prevent the closedChannel list from exceeding the max size
+            // by removing the oldest element
+            if (closedChannelIds.Count >= closed_channels_max_size)
+            {
+                closedChannelIds.RemoveAt(0);
+            }
+
+            // For PM channels, we store the user ID; else, we store the channel id
+            closedChannelIds.Add(channel.Type == ChannelType.PM ? channel.Users.First().Id : channel.Id);
+
             if (channel.Joined.Value)
             {
                 api.Queue(new LeaveChannelRequest(channel));
                 channel.Joined.Value = false;
             }
         }
 
+        /// <summary>
+        /// Opens the most recently closed channel that has not
+        /// already been reopened
+        /// Works similarly to reopening the last closed tab on a web browser.
+        /// </summary>
+        public void JoinLastClosedChannel()
+        {
+            if (closedChannelIds.Count <= 0)
+            {
+                return;
+            }
+
+            // This nested loop could be eliminated if a check was added so that
+            // when the code opens a channel it removes from the closedChannel list.
+            // However, this would require adding an O(|closeChannels|) work operation
+            // every time the user joins a channel, which would make joining a channel
+            // slower. I wanted to centralize all major slowdowns so they
+            // can only occur if the user actually decides to use this feature.
+            while (closedChannelIds.Count != 0)
+            {
+                long lastClosedChannelId = closedChannelIds.Last();
+                closedChannelIds.RemoveAt(closedChannelIds.Count - 1);
+
+                bool lookupCondition(Channel ch) => ch.Type == ChannelType.PM
+                    ? ch.Users.Any(u => u.Id == lastClosedChannelId)
+                    : ch.Id == lastClosedChannelId;
+
+                // If the user hasn't already joined the channel, try to join it
+                if (joinedChannels.FirstOrDefault(lookupCondition) == null)
+                {
+                    Channel lastClosedChannel = AvailableChannels.FirstOrDefault(lookupCondition);
+
+                    if (lastClosedChannel != null)
+                    {
+                        CurrentChannel.Value = JoinChannel(lastClosedChannel);
+                    }
+                    else
+                    {
+                        // Try to get User to open PM chat
+                        var req = new GetUserRequest(lastClosedChannelId);
+                        req.Success += user => CurrentChannel.Value = JoinChannel(new Channel(user));
+                        api.Queue(req);
```

Comment:
This will unnecessarily re-fetch the user every time. The proper solution for this is to use `UserLookupCache` which will 100% have the user locally fetched already.